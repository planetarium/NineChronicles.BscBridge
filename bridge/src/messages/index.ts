import { ChatPostMessageArguments } from "@slack/web-api";
import { ForceOmit } from "../types/force-omit";

export interface Message {
    render(): ForceOmit<ChatPostMessageArguments, "channel">;
}
