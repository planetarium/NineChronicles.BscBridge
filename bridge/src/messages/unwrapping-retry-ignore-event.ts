import { ChatPostMessageArguments } from "@slack/web-api";
import { Message } from ".";
import { ForceOmit } from "../types/force-omit";
import { TxId } from "../types/txid";

export class UnwrappingRetryIgnoreEvent implements Message {
  private readonly _txId: string;

  constructor(txId: TxId) {
    this._txId = txId;
  }

  render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
    return {
      text: "wNCG → NCG event already seems executed so it skipped.",
      attachments: [
        {
          author_name: "Bridge Warning",
          color: "#ffcc00",
          fields: [
            {
              title: "Binance Smart Chain network transaction id",
              value: this._txId,
            },
          ],
          fallback: `wNCG → NCG event already seems executed so it skipped.`,
        },
      ],
    };
  }
}
