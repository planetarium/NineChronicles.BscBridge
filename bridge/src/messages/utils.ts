import { join } from "path";
import { URL } from "url";

export function combineNcExplorerUrl(
  explorerUrl: string,
  ncscanUrl: string | undefined,
  useNcscan: boolean,
  txId: string
) {
  if (useNcscan) {
    if (ncscanUrl === undefined) {
      throw new Error("ncscanUrl is undefined");
    }

    return combineUrl(ncscanUrl, `/tx/${txId}`);
  } else {
    return combineUrl(explorerUrl, "/transaction", txId);
  }
}

export function combineUrl(base: string, path: string, query?: string): string {
  const url = new URL(base);
  url.pathname = join(url.pathname, path);
  url.search = query || "";
  return url.toString();
}
