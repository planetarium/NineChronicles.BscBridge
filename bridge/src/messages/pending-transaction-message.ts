import { ChatPostMessageArguments } from "@slack/web-api";
import { Message } from ".";
import { ForceOmit } from "../types/force-omit";
import { ExchangeHistory } from "../interfaces/exchange-history-store";
import { combineUrl } from "./utils";
import { MultiPlanetary } from "../multi-planetary";

export class PendingTransactionMessage implements Message {
  private readonly _bscScanUrl: string;
  private readonly _multiPlanetary: MultiPlanetary;

  constructor(
    private readonly transactions: ExchangeHistory[],
    private readonly multiPlanetary: MultiPlanetary,
    bscScanUrl: string = process.env.BSCSCAN_URL || "https://bscscan.com"
  ) {
    this._bscScanUrl = bscScanUrl;
    this._multiPlanetary = multiPlanetary;
  }

  render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
    if (this.transactions.length !== 0) {
      console.log("Pending Transactions : ", this.transactions);
      return {
        text: `${this.transactions.length} Pending Transactions Found`,
        attachments: this.transactions.map((tx) => ({
          author_name: "[BSC] wNCG â†’ NCG pending event",
          color: "#ff0033",
          fields: [
            {
              title: "BSC transaction",
              value: combineUrl(this._bscScanUrl, `/tx/${tx.tx_id}`),
            },
            {
              title: "Planet Name",
              value: this._multiPlanetary.getRequestPlanetName(tx.recipient),
            },
            {
              title: "Sender(BSC)",
              value: tx.sender,
            },
            {
              title: "Recipient(9c)",
              value: tx.recipient,
            },
            {
              title: "Amount",
              value: tx.amount.toString(),
            },
            {
              title: "Timestamp",
              value: tx.timestamp,
            },
          ],
        })),
      };
    }

    return {
      text: "No pending transactions",
      attachments: [
        {
          author_name: "BSC Bridge Restarted",
          color: "#ffcc00",
        },
      ],
    };
  }
}
