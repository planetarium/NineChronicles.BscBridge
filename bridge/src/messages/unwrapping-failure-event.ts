import { ChatPostMessageArguments } from "@slack/web-api";
import { Message } from ".";
import { TxId } from "../types/txid";
import { Address } from "../types/address";
import { ForceOmit } from "../types/force-omit";
import { combineUrl } from "./utils";

export class UnwrappingFailureEvent implements Message {
    private readonly _url: string;
    private readonly _sender: Address;
    private readonly _recipient: Address;
    private readonly _txId: TxId;
    private readonly _amount: string;
    private readonly _error: string;
    private readonly _planetName: string;
    private readonly _subscribers: string;

    constructor(
        url: string,
        sender: Address,
        recipient: Address,
        amount: string,
        txId: TxId,
        error: string,
        planetName: string,
        subscribers: string
    ) {
        this._url = url;
        this._sender = sender;
        this._recipient = recipient;
        this._amount = amount;
        this._txId = txId;
        this._error = error;
        this._planetName = planetName;
        this._subscribers = subscribers;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        return {
            text: `[BSC] wNCG → NCG event failed. ${this._subscribers}`,
            attachments: [
                {
                    author_name: "Bridge Error",
                    color: "#ff0033",
                    fields: [
                        {
                            title: "BSC transaction",
                            value: combineUrl(this._url, `/tx/${this._txId}`),
                        },
                        {
                            title: "sender (bsc)",
                            value: this._sender,
                        },
                        {
                            title: "recipient (NineChronicles)",
                            value: this._recipient,
                        },
                        {
                            title: "amount",
                            value: this._amount,
                        },
                        {
                            title: "Planet Name",
                            value: this._planetName,
                        },
                        {
                            title: "error",
                            value: this._error,
                        },
                    ],
                    fallback: `wNCG ${this._sender} → NCG ${this._recipient} failed`,
                },
            ],
        };
    }
}
