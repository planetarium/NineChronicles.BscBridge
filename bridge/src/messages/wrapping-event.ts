import { ChatPostMessageArguments } from "@slack/web-api";
import { Message } from ".";
import { TxId } from "../types/txid";
import { URL } from "url";
import { join, resolve } from "path";
import { ForceOmit } from "../types/force-omit";
import { combineNcExplorerUrl, combineUrl } from "./utils";

export abstract class WrappingEvent implements Message {
    private readonly _explorerUrl: string;
    private readonly _ncscanUrl: string | undefined;
    protected readonly _useNcscan: boolean;
    private readonly _etherscanUrl: string;

    constructor(
        explorerUrl: string,
        ncscanUrl: string | undefined,
        useNcscan: boolean,
        etherscanUrl: string
    ) {
        this._explorerUrl = explorerUrl;
        this._ncscanUrl = ncscanUrl;
        this._useNcscan = useNcscan;
        this._etherscanUrl = etherscanUrl;
    }

    abstract render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel">;

    protected toExplorerUrl(txId: TxId): string {
        return combineNcExplorerUrl(
            this._explorerUrl,
            this._ncscanUrl,
            this._useNcscan,
            txId
        );
    }

    protected toEtherscanUrl(transactionHash: string): string {
        return combineUrl(this._etherscanUrl, `/tx/${transactionHash}`);
    }
}
