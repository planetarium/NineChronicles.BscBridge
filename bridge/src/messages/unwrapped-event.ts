import { ChatPostMessageArguments } from "@slack/web-api";
import { WrappingEvent } from "./wrapping-event";
import { Address } from "../types/address";
import { TxId } from "../types/txid";
import { ForceOmit } from "../types/force-omit";

export class UnwrappedEvent extends WrappingEvent {
    private readonly _sender: Address;
    private readonly _recipient: Address;
    private readonly _nineChroniclesTxId: TxId;
    private readonly _ethereumTransactionHash: string;
    private readonly _amount: string;
    private readonly _isMultiPlanetRequestType: boolean;
    private readonly _planetName: string;

    constructor(
        explorerUrl: string,
        ncscanUrl: string | undefined,
        useNcscan: boolean,
        etherscanUrl: string,
        sender: Address,
        recipient: Address,
        amount: string,
        nineChroniclesTxId: TxId,
        ethereumTransactionHash: string,
        isMultiPlanetRequestType: boolean,
        planetName: string
    ) {
        super(explorerUrl, ncscanUrl, useNcscan, etherscanUrl);

        this._sender = sender;
        this._recipient = recipient;
        this._amount = amount;
        this._nineChroniclesTxId = nineChroniclesTxId;
        this._ethereumTransactionHash = ethereumTransactionHash;
        this._planetName = planetName;
        this._isMultiPlanetRequestType = isMultiPlanetRequestType;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        const message = {
            text: "[BSC] wNCG → NCG event occurred.",
            attachments: [
                {
                    author_name: "Bridge Event",
                    color: "#f5b342",
                    fields: [
                        {
                            title: "9c network transaction",
                            value: this.toExplorerUrl(this._nineChroniclesTxId),
                        },
                        {
                            title: "Binance Smart Chain network transaction",
                            value: this.toEtherscanUrl(
                                this._ethereumTransactionHash
                            ),
                        },
                        {
                            title: "sender (bsc)",
                            value: this._sender,
                        },
                        {
                            title: "recipient (NineChronicles)",
                            value: this._recipient,
                        },
                        {
                            title: "amount",
                            value: this._amount,
                        },
                        {
                            title: "Planet-Name",
                            value: this._planetName,
                        },
                        {
                            title: "Network",
                            value: 'Binance Smart Chain'
                        },
                    ],
                    fallback: `wNCG ${this._sender} → NCG ${this._recipient}`,
                },
            ],
        };

        if (!this._isMultiPlanetRequestType) {
            message.attachments[0].fields.push({
                title: "Not Multi-Planet Request Type",
                value: "This Transfer Request is not a multi-planet request type.",
            });
        }

        return message;
    }
}
