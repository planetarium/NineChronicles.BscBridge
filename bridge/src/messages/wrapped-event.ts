import { ChatPostMessageArguments } from "@slack/web-api";
import { WrappingEvent } from "./wrapping-event";
import { Address } from "../types/address";
import { TxId } from "../types/txid";
import Decimal from "decimal.js";
import { ForceOmit } from "../types/force-omit";

export class WrappedEvent extends WrappingEvent {
    private readonly _sender: Address;
    private readonly _recipient: Address;
    private readonly _nineChroniclesTxId: TxId;
    private readonly _ethereumTransactionHash: string;
    private readonly _fee: Decimal;
    private readonly _exchangeAmount: string;
    private readonly _refundAmount: string | null;
    private readonly _refundTxId: string | null;
    private readonly _isWhitelistEvent: boolean;
    private readonly _description: string | undefined;

    constructor(
        explorerUrl: string,
        ncscanUrl: string | undefined,
        useNcscan: boolean,
        etherscanUrl: string,
        sender: Address,
        recipient: Address,
        exchangeAmount: string,
        nineChroniclesTxId: TxId,
        ethereumTransactionHash: string,
        fee: Decimal,
        refundAmount: string | null,
        refundTxId: TxId | null,
        isWhitelistEvent: boolean,
        description: string | undefined
    ) {
        super(explorerUrl, ncscanUrl, useNcscan, etherscanUrl);

        this._sender = sender;
        this._recipient = recipient;
        this._exchangeAmount = exchangeAmount;
        this._nineChroniclesTxId = nineChroniclesTxId;
        this._ethereumTransactionHash = ethereumTransactionHash;
        this._fee = fee;
        this._refundAmount = refundAmount;
        this._refundTxId = refundTxId;
        this._isWhitelistEvent = isWhitelistEvent;
        this._description = description;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        const refundFields =
            this._refundAmount !== null && this._refundTxId !== null
                ? [
                      {
                          title: "refund amount",
                          value: this._refundAmount,
                      },
                      {
                          title: "refund transaction",
                          value: this.toExplorerUrl(this._refundTxId),
                      },
                  ]
                : [];

        let text = "NCG → wNCG event occurred.";
        if (this._isWhitelistEvent) text += " (Whitelist Transfer) <!here>";

        const message = {
            text,
            attachments: [
                {
                    author_name: "Bridge Event",
                    color: !this._isWhitelistEvent ? "#42f5aa" : "#b547f5",
                    fields: [
                        {
                            title: "9c network transaction",
                            value: this.toExplorerUrl(this._nineChroniclesTxId),
                        },
                        {
                            title: "Ethereum network transaction",
                            value: this.toEtherscanUrl(
                                this._ethereumTransactionHash
                            ),
                        },
                        {
                            title: "sender (NineChronicles)",
                            value: this._sender,
                        },
                        {
                            title: "recipient (Ethereum)",
                            value: this._recipient,
                        },
                        {
                            title: "amount",
                            value: this._exchangeAmount,
                        },
                        {
                            title: "fee",
                            value: this._fee.toString(),
                        },
                        ...refundFields,
                    ],
                    fallback: `NCG ${this._sender} → wNCG ${this._recipient}`,
                },
            ],
        };

        if (this._description) {
            message.attachments[0].fields.push({
                title: "description",
                value: this._description,
            });
        }

        return message;
    }
}
