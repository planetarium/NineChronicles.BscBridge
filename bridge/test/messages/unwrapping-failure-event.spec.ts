import { UnwrappingFailureEvent } from "../../src/messages/unwrapping-failure-event";

describe("UnwrappingFailureEvent", () => {
  describe("render", () => {
    it("snapshot", () => {
      const ETHERSCAN_URL = "https://ropsten.etherscan.io";
      const SENDER = "0xDac65eCE9CB3E7a538773e08DE31F973233F064f";
      const RECIPIENT = "0xCbfC996ad185c61a031f40CeeE80a055e6D83005";
      const AMOUNT = "100";
      const ETHEREUM_TRANSACTION_HASH =
        "0x9360cd40682a91a71f0afbfac3dd381866cdb319dc01c13531dfe648f8a28bc7";
      expect(
        new UnwrappingFailureEvent(
          ETHERSCAN_URL,
          SENDER,
          RECIPIENT,
          AMOUNT,
          ETHEREUM_TRANSACTION_HASH,
          String(new Error("error")),
          "odin",
          "@gamefi-be"
        ).render()
      ).toMatchSnapshot();
    });
  });
});
