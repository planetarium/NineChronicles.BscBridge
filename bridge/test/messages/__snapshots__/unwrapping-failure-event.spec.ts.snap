// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnwrappingFailureEvent render snapshot 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_name": "Bridge Error",
      "color": "#ff0033",
      "fallback": "wNCG 0xDac65eCE9CB3E7a538773e08DE31F973233F064f → NCG 0xCbfC996ad185c61a031f40CeeE80a055e6D83005 failed",
      "fields": Array [
        Object {
          "title": "Ethereum transaction",
          "value": "https://ropsten.etherscan.io/tx/0x9360cd40682a91a71f0afbfac3dd381866cdb319dc01c13531dfe648f8a28bc7",
        },
        Object {
          "title": "sender (Ethereum)",
          "value": "0xDac65eCE9CB3E7a538773e08DE31F973233F064f",
        },
        Object {
          "title": "recipient (NineChronicles)",
          "value": "0xCbfC996ad185c61a031f40CeeE80a055e6D83005",
        },
        Object {
          "title": "amount",
          "value": "100",
        },
        Object {
          "title": "Planet Name",
          "value": "odin",
        },
        Object {
          "title": "error",
          "value": "Error: error",
        },
      ],
    },
  ],
  "text": "wNCG → NCG event failed. @gamefi-be",
}
`;
